Interface extension and `final impl` update (#5337)
carbon-language/carbon-lang: proposals/p5337.md

-   [Alternatives considered](#alternatives-considered)
    -   [Allow overlap between a non-final and final impl but only if no queries pick the non-final on the overlap](#allow-overlap-between-a-non-final-and-final-impl-but-only-if-no-queries-pick-the-non-final-on-the-overlap)
    -   [Forbid overlap between final impls](#forbid-overlap-between-final-impls)
    -   [Prioritize between final impls using type structure](#prioritize-between-final-impls-using-type-structure)
    -   [Allow mixing final and non-final impls in the same `match_first` block](#allow-mixing-final-and-non-final-impls-in-the-same-match_first-block)
    -   [No default `Self` in `require Self impls I`](#no-default-self-in-require-self-impls-i)

<!-- tocstop -->

## Abstract

We make 5 changes:

-   Allow `require Self impls I` in an `interface` or `constraint` scope to omit
    the `Self`, so it can be written `require impls I`.
-   Rename `extend I` to `extend require impls I` in an `interface` or
    `constraint` scope.
-   Define `extend impl as I` and `extend final impl as I` in an `interface`
    scope to copy the members of `I` and define an `impl` of `I` in terms of the
    extending interface.

> -   Allow an `impl` to overlap a final `impl` as long as it isn't subsumed by

Reviewer: In general each time this proposal writes "`impl`" it would be nice to ensure it either says:
- "non-final `impl`", or
- "`final impl`"

And if it means both, perhaps to treat them as disjoint things, to avoid ambiguity: "non-final `impl` or `final impl`"?

Otherwise I think right now it's sometimes saying `impl` as in "non-final" and sometimes as in "final or non-final". At the least maybe do a pass to ensure `impl` is always qualified with "non-final" if that's what it means?

Also sometimes "final `impl`" and sometimes "`final impl`", so it would be nice to be consistent.

Author: I did a search for references to "impl" and tried to make sure the text was clear. I ended up making some other clarifications due to rereading the text.

The `final impl` thing looks inconsistent at first, but I was careful to only say `final impl` when I meant "an `impl` declaration with a `final` modifier". A "final `impl`" is one declared using `final impl` or an `impl` that is named in a `final match_first`. But that definition comes pretty late in the text.

    introduced `require Self impls I` in an `interface` or `constraint` scope,
    replacing the previous spelling of `impl as I` from
    [proposal #553](https://github.com/carbon-language/carbon-lang/pull/553).
-   [Proposal #2868](https://github.com/carbon-language/carbon-lang/pull/2868)
    allowed an `impl` to overlap a `final impl` as long as agreed on the
    overlap.

## Proposal

First, since we are making the `require Self impls I` usage more common in
`interface` and `constraint` scopes, we allow the `Self` to be omitted. So:

-   `require Self impls I` ✅ allowed, same meaning as before
-   `require impls I` ✅ allowed, same as `require Self impls I`
-   `require T impls I` ✅ allowed, same meaning as before

Next, since we are creating another way that an `interface` can `extend` another
interface, we change the existing usage of `extend` from being an introducer of
its own declaration, to a modifier on a `require` declaration. In this case, the
only type that can be constrained is `Self`:

-   `extend require Self impls I` ✅ allowed
-   `extend require impls I` ✅ allowed, same as `extend require Self impls I`
-   `require T impls I` ❌ forbidden

> This echos the similar restriction on `extend impl` in `class` scope.

Reviewer: But `extend impl as` does not allow `extend impl Self as`, right? Why make this one different?

Edit: To clarify, why is `Self` allowed in the `extend` case here?

Author: You are right, we should be consistent and forbid `Self` here. The same reasoning applies.

`extend require impls I` after this proposal has the same meaning as `extend I`
did before this proposal, so:

```carbon
interface A {
  fn F();
}

interface B {
  extend require impls A;
  fn G();
}
```

is equivalent to:

```carbon
interface A {
  fn F();
}

interface B {
  require impls A;
  alias F = A.F;
  fn G();
}
```

We then add another way to use `extend` as a modifier in an `interface` scope.
We define the meaning of `extend impl as I;` in an `interface J` scope to do two
things:

-   Copy the members of `I` to form new members of `J`.
-   Define a blanket `impl` that anything that implements `J` also implements
    `I`, by forwarding to the corresponding members of `J`.

These semantics are intended to be similar to `extend impl` in a `class`, except

> that the definition of the `impl` is generated automatically (and `impl` without

Reviewer: ```suggestion
that the definition of the blanket `impl` is generated automatically (and `impl` without
```

  fn H();
}
```

The definition of `interface J` is equivalent to:

```carbon
interface J {
  let T:! type;
  fn F();
  fn G();
  fn H();
}
impl forall [U:! J] U as I {
  let T:! type = U.(J.T);
  fn F() = U.(J.F);
  fn G() = U.(J.G);
}
```

The keyword `final` can be added between `extend` and `impl` to make the
generated `impl` definition `final`. For example, the
`final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T)` from
[the "Problem" section](#problem) would be generated from this:

```carbon
interface ImplicitAs(T:! type) {
  extend final impl as As(T);
}
```

To allow an `impl` to overlap a final `impl`, we say the final `impl` is
prioritized over a non-final `impl` anytime they overlap. This addresses a
[problem](#problem) caused by the fact that we don't support the negative
constraints that would allow a developer to avoid the overlap -- and the only
sound choice is to prefer the final `impl`. However, if an `impl` can never be
used because it is completely subsumed by a final `impl`, that is an error.

To allow multiple interfaces extending the same interface, we also want to
support a way to declare overlapping final impls. In this case, list the impls

> in a `final match_first` block instead of putting `final` on the `impl`

Reviewer: I feel uncomfortable about the match_first affecting more than priority here, and mutating the final-ness of the impl still.

I know we talked about how to implement it, and it's fine from a toolchain perspective, by making a copy of the impl but as final. But as a code author I think this is going to be a sharp corner in the language, where adding `final` in one place has an effect on a bunch of impls that are not directly present with that final keyword in the code.

I would much prefer if we could say that `final match_first` can only contain final impls, or names of final impls. Then if you wanted to change a match_first to final, you'd get errors and see exactly which other impls are becoming final and mark them accordingly.

I guess this doesn't work because you end up with overlapping final impls before you get to the match_first. Maybe this is just something to discuss in the alternatives... but this does feel like a sharp edge to me.

```carbon
interface A {
  let T:! type;
  fn F() -> T;
}

interface B {
  let T:! type;
  // Okay
  extend impl as A where .T = T*;
}

class C {
  impl as B {
    // `T` here is `B.T`
    where T = i32;
    // `F` here is `B.F` and `A.F`.
    fn F() -> i32*;
  }
}
```

This is equivalent to:

```carbon
interface A {
  let T:! type;
  fn F() -> T;
}

interface B {
  let T:! type;
  // Skip A.T due to conflict.
  fn F() -> T*;
}

impl forall [U:! B] U as A where .T = .(B.T)* {
  fn F() -> T = U.(B.F);
}

class C {
  impl as B {
    where T = i32;
    fn F() -> i32*;
  }
}
```

This was used in the `OutputForwardIterator` example in
[the "Proposal" section](#proposal).

Name conflicts are not expected to generally be a problem in practice since this
form of extension often requires the two interfaces to be defined in the same
file.

### Final `impl` priority

With the design prior to this proposal, a final `impl` can only overlap another
`impl` if they have the same definition on their overlap. This means that if
there is a `final` blanket impl of an interface, as is generated by
`extend final impl as` in an interface, then other impls of that interface may

> overly restricted. This example:

Reviewer: "may _be_ overly restricted"?

```carbon
package Core;

interface As(T:! type) {
  fn Convert[self: Self]() -> T;
}

interface ImplicitAs(T:! type) {
  extend final impl as As(T);
}
```

is equivalent to:

```carbon
package Core;

interface As(T:! type) {
  fn Convert[self: Self]() -> T;
}

interface ImplicitAs(T:! type) {
  fn Convert[self: Self]() -> T;
}

final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T) {
  fn Convert[self: Self]() -> T = U.(ImplicitAs(T).Convert);
}
```

Any other implementation of `As(T)` is going to overlap unless we could
establish that the type does not implement `ImplicitAs(T)`. However, in a
generic context, we don't support (and don't intend to support) that sort of
negative constraint. So there would be no way to prove this `impl` of `As(T)`

> doesn't overlap the `final impl` of `As(T)`:

Reviewer: ```suggestion
Any other `impl` of `As(T)` is going to overlap unless we could
establish that the type does not implement `ImplicitAs(T)`. However, in a
generic context, we don't support (and don't intend to support) that sort of
negative constraint. So there would be no way to prove the following `impl` of
`As(T)` doesn't overlap the `final impl` of `As(T)`:
```

```carbon
import Core;

class X(T:! type) {
  impl as As(T);
}
```

In particular, an unrelated library defining `MyType` could define
`impl X(MyType) as ImplicitAs(MyType)`, which would create a conflict between
the `impl` in `X(T)` and the `final impl` associated with `ImplicitAs(T)`.

This motivates a change to the rules: we want to allow an `impl` to overlap with
a final `impl`. The question is what to do for queries in the overlap.

Now consider this second example:

```carbon
import Core;

class Array(Element:! type) { ... }

// Impl 1
impl forall [T:! type, U:! As(T)]
    Array(U) as As(Array(T)) { ... }

// Impl 2
impl forall [T:! type, U:! ImplicitAs(T)]

>     Array(U) as ImplicitAs(Array(T)) { ... }

Reviewer: What if the library wants to write these as final so that it can ensure it knows the implementation of `As` for `Array`? I think that would be diagnosed as an error right? This feels like an example that I might really not want to allow to be specialized, and if not, we could come up with another one that does?

Using `final` as a priority marker feels like it's suffering from being essentially one bit of priority, which works for priority between interface and impl, but then is not enough when there's multiple levels of impls possible?

If match_first wasn't the way to make an interface final, then maybe we could match_first between this final impl and the one in the `As` interface, in order to allow writing `final` here, but I guess that's ruled out right now too.

Author: Technically, these can be written as `final` under the rules of this proposal, but the blanket final `impl` in `Core` will be prioritized.

```

In this example, there is overlap between the definition of `As` provided by
"Impl 1", and the definition of `As` provided by the `final impl` from
`ImplicitAs`. For example, if `i32 impls ImplicitAs(i64)`, then "Impl 2" defines

> how `Array(i32) impls ImplicitAs(Array(i64))`, and the `final impl` gives us a

Reviewer: I don't think there's any `As` with a `final extends` written down in this proposal yet at this point, certainly not nearby in a way that's easy to connect from this example as being the one true `As`. Maybe the example should include it?

In general I found this paragraph overly difficult to read/pull apart, maybe it can be improved in some way at the same time. I thought about maybe bulleted steps to get from the example code to the overlap error? Not sure what would look best for you.

Maybe instead of `X impls Y` it could be `impl X as Y` or `final impl X as Y`? Writing it in terms of the impls decls might help instead of as constraints which are not part of the examples.

I had trouble following how we get the overlap in the end. I guess it's saying the blanket impls here each provide the explicit Array(U) -> Array(T) conversion if an implicit U -> T conversion exists?

Author: Step 1: impl 1 provides Array(U) as As(Array(T))
Step 2: impl 2 provides Array(U) as ImplicitAs(Array(T))
Step 3: the ImplicitAs interface has an extend final impl of As
Step 4: Step 2 and 3 combine to give another impl of As(Array(T)) that conflicts with the impl from step 1

Good to note: we want both impl 1 and impl 2. Impl 1 is for types T that impl As but not ImplicitAs. Impl 2 is for types that impl ImplicitAs.

Author: Rewrote this paragraph.

type are only selected if they match and no final `impl` definitions in the
library with the interface do (in the case those are different files).

> _**Concern**_: This creates a situation where a final `impl` defined in the
> file with the type may not be selected for some queries that it matches, which
> may be surprising.

#### Overlapping `final` impls

If there is any other overlap between final `impl` definitions, it must be
within the same file. We have two tools here we could use to select between the
candidates: type structure and
[`impl_priority`/`match_first` blocks](/docs/design/generics/details.md#prioritization-rule).
(We will use the placeholder term of "`match_first` blocks" to refer to these
without asserting how they will be spelled.)

The decision here is to only use `match_first` blocks, not type structure to
choose between overlapping final `impl` definitions in the same file, for a few
reasons:

-   It would be surprising to prefer something with a specified type over a
    blanket `impl` when they are in the same file, as type structure would
    dictate, since we are saying the blanket `impl` will have priority when they

>     are in different files.

Reviewer: But we said the blanket one has priority only because it's final, and I think the current design prevents the specialized one from being final (which I think is a problem but mentioned above...). In this case to have priority the specialization with the more specified type would have to be final so I am not sure this is surprising here.

  extend impl as A(i32);
  // Produces impl with name `B.(as A(i32))`.
}

interface C1 {
  let U:! type;
  extend impl as A(U);
  // Produces impl with name `C1.(as A(U))`.
}

interface C2 {
  let U:! type;
  // Only difference from C1 is using `Self.U` instead of `U`.
  extend impl as A(Self.U);
  // Produces impl with name `C2.(as A(Self.U))`.
}

interface D(V:! type) {
  extend impl as A(V);
  // Produces impl with name `D(V:! type).(as A(V))`.
}

interface E {}
impl forall [W:! E] W as A(W) { ... }

class F {}
impl F as A(F) { ... }

// Placeholder impl priority syntax
match_first {
  impl B.(as A(i32));

  impl C1.(as A(U));
  impl C2.(as A(Self.U));
  // Can't write `impl C1.(as A(Self.U))` or `impl C2.(as A(U))`.

  impl D(V:! type).(as A(V));

  impl forall [W:! E] W as A(W);

  impl F as A(F);
}
```

Note how the expression after the `as` has to match syntactically between the
two declarations.

Without the `impl` prioritization, these would be conflicting implementations of
the interface `A`. Note that by the
[overlapping `final` impl rules](#overlapping-final-impls), at most one
interface can `final extend impl as` a given interface. `final match_first` must

> be used instead of multiple, so that there is an explicit prioritization between

Reviewer: ```suggestion
be used instead of multiple `final impl`s, so that there is an explicit prioritization between
```

[proposal #3763: Matching redeclarations](https://github.com/carbon-language/carbon-lang/pull/3763),
the same sequence of tokens (between the `interface` and `{`) is used to name
the interface, allowing a syntactic match. For example:

```
interface Z(T:! type) { ... }

interface Y(T:! type) {
  extend impl as Z(T*);
}
// ...

final match_first {
  impl Y(T:! type).(as Z(T*));
  // ...
}
```

## Rationale

This proposal's end goal is to support a form of interface extension that
matches user's expectations, to aid the
["Code that is easy to read, understand, and write" goal](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write).
This goal also informed the choice to be explicit about prioritizing overlapping
final impls.

The choice to not treat an `impl` as final when it overlaps another final impl
until we see how to choose between them on the overlap is in accord with the
[information accumulation principle](/docs/project/principles/information_accumulation.md).

## Alternatives considered

### Allow overlap between a non-final and final impl but only if no queries pick the non-final on the overlap

This question was considered in the

> ["final `impl` priority" section](#final-impl-priority).

Reviewer: Is it still? I don't see this discussed

Author: Look for the text starting with "The question is what to do for queries in the overlap." and in particular "And we don't want to give an error here since:...".

    -   [Forbid overlap between final impls](#forbid-overlap-between-final-impls)
    -   [Prioritize between final impls using type structure](#prioritize-between-final-impls-using-type-structure)
    -   [Allow mixing final and non-final impls in the same `match_first` block](#allow-mixing-final-and-non-final-impls-in-the-same-match_first-block)
    -   [No default `Self` in `require Self impls I`](#no-default-self-in-require-self-impls-i)

<!-- tocstop -->

## Abstract

We make 5 changes:

-   Allow `require Self impls I` in an `interface` or `constraint` scope to omit
    the `Self`, so it can be written `require impls I`.
-   Rename `extend I` to `extend require impls I` in an `interface` or
    `constraint` scope.
-   Define `extend impl as I` and `extend final impl as I` in an `interface`
    scope to copy the members of `I` and define an `impl` of `I` in terms of the
    extending interface.
-   Allow a non-final `impl` to overlap a final `impl` as long as it isn't
    subsumed by the final `impl`. The final `impl` will be given priority on the
    overlap.

> -   Allow `final` on a `match_first` block, used to declare overlapping final

Reviewer: This does not mention that final impls can overlap if one comes from the root self type library and one from the interface library (and they are different libraries). Either this should be generalized or made to include both cases?

Author: Done.

}
```

The definition of `interface J` is equivalent to:

```carbon
interface J {
  let T:! type;
  fn F();
  fn G();
  fn H();
}
impl forall [U:! J] U as I {
  let T:! type = U.(J.T);
  fn F() = U.(J.F);
  fn G() = U.(J.G);
}
```

The keyword `final` can be added between `extend` and `impl` to make the
generated `impl` definition `final`. For example, the
`final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T)` from
[the "Problem" section](#problem) would be generated from this:

```carbon
interface ImplicitAs(T:! type) {
  extend final impl as As(T);
}
```

To allow a non-final `impl` to overlap a final `impl`, we say the final `impl`
is prioritized over a non-final `impl` anytime they overlap. This addresses a
[problem](#problem) caused by the fact that we don't support the negative
constraints that would allow a developer to avoid the overlap -- and the only
sound choice is to prefer the final `impl`. However, if a non-final `impl` can
never be used because it is completely subsumed by a final `impl`, that is an
error.

To allow multiple interfaces extending the same interface, we also want to
support a way to declare overlapping final `impl`s. In this case, list the

> `impl` names in a `final match_first` block instead of putting `final` on the

Reviewer: This does not mention that final impls can overlap if one comes from the root self type library and one from the interface library (and they are different libraries).

Author: Added text.

}
```

Any other `impl` of `As(T)` is going to overlap unless we could establish that
the type does not implement `ImplicitAs(T)`. However, in a generic context, we
don't support (and don't intend to support) that sort of negative constraint. So
there would be no way to prove the following `impl` of `As(T)` doesn't overlap
the `final impl` of `As(T)`:

```carbon
import Core;

class X(T:! type) {
  impl as As(T);
}
```

In particular, an unrelated library defining `MyType` could define
`impl X(MyType) as ImplicitAs(MyType)`, which would create a conflict between
the `impl` in `X(T)` and the `final impl` associated with `ImplicitAs(T)`.

This motivates a change to the rules: we want to allow a non-final `impl` to
overlap with a final `impl`. The question is what to do for queries in the
overlap.

Now consider this second example:

```carbon
import Core;

class Array(Element:! type) { ... }

// Impl 1
impl forall [T:! type, U:! As(T)]
    Array(U) as As(Array(T)) { ... }

// Impl 2
impl forall [T:! type, U:! ImplicitAs(T)]
    Array(U) as ImplicitAs(Array(T)) { ... }
```

In this example, there is overlap between the definition of `As` provided by

> "Impl 1", and the definition of `As` provided by the `final impl` from

Author: Reference the definition of `package Core` from the beginning of this section and link to it `[Final `impl` priority](#final-impl-priority)`

then `i32 impls As(i64)` and "Impl 1" would provide another definition of
`Array(i32) impls As(Array(i64))` that is higher priority by
[the type-structure rule](/docs/design/generics/details.md#overlap-rule),
creating a conflict.

Any generic code that sees the final `impl` can assume it applies whenever it is
used. If we don't want to forbid this overlap situation (either when defining a
non-final `impl` with a more-specific type structure or when doing an `impl`
query in the overlap that selects the non-final one), the only sound choice is
to use the final `impl`. And we don't want to give an error here since:

-   It would reduce expressiveness, since there isn't a good workaround to
    prevent the overlap.
-   It would reduce the ability to compose libraries.
-   We want to support examples like this interaction between `Array`, `As`, and
    `ImplicitAs`.

Note that prioritizing the final `impl` over the non-final `impl` (even when the
non-final `impl` has a preferred type structure) produces the desired result in
the example above. By writing `final` we ensure that `Convert` in `ImplicitAs`
always matches `Convert` in `As`, when they are both defined.

> The next question is what to do if there are two overlapping final `impl`s. The

Author: Introduce a section break here

type and the one defining the root interface. The only way to have the final
`impl`s in different files is if there is a
[blanket impl](/docs/design/generics/details.md#blanket-impl-declarations) of
the interface in the file defining the interface and the other in the file
defining the type. This latter file will have to transitively depend on the file
defining the interface in order to write its final `impl` definition, and so
will be able to see any `final` blanket impls.

This means there may be files that will only see the final `impl` definitions in
the same file as the interface definition. Those final `impl` definitions will
be assumed to apply whenever they match, and so they must be prioritized over
any other overlapping final `impl` defined in the file with the type. We
therefore adopt the rule that final `impl` definitions in the library with the
type are only selected if they match and no final `impl` definitions in the
library with the interface do (in the case those are different files).

> _**Concern**_: This creates a situation where a final `impl` defined in the
> file with the type may not be selected for some queries that it matches, which
> may be surprising.

> #### Overlapping `final` impls

Author: This is specifically the case of overlap within a single library/file.

> If there is any other overlap between final `impl` definitions, it must be

Reviewer: I think that we should consider saying that final impls just use the inverse type structure ordering as non-final impls. This would still select impls in the interface file over those in the self-type file, but it would avoid a problem of the code changing meaning (and breaking) when you join files together.

It's common in C++ to concatenate files together, for example when submitting a minimized repro of a clang crash. Or to refactor code to move it between files.

Under the rules here, combining the files where the interface and specialized type are would break then build and you'd need to combine the impls together into a match_first statement manually. If we said the type structure is used but in the reverse priority for `final`, then it would continue to work wherever you moved the self-type and its final impl.

Also there was an argument made that match_first reduces the cognitive work of knowing how final impls will be prioritized but I think that doesn't hold when the final impl is in another file anyways - and that will be the more common case too. So I believe it's a simpler rule to teach that we always use type structure but the priority is inverted for `final`. This would make match_first more of a power tool to reach for when you need to do something special.

Author: Discussion in [#generics-and-templates on Discord](https://discord.com/channels/655572317891461132/941071822756143115/1367982721141440734)

-   Last priority: non-final `impl` definitions

In the case that there are multiple matching final `impl`s defined in the same
file, they overlap so they must be in a single `final match_first` block. In
that case, pick the first one listed in the `final match_first` block.

If no final `impl` declarations match the query, we fall back to the original
rules:

-   The non-final `impl` declarations matching the query with the most preferred
    type structure must be in the same non-final `match_first` block, or a
    single non-final `impl` declaration not in a `match_first` can match.
    -   In the latter case, that `impl` is selected. Alternatively, `impl`
        declarations not in any `match_first` block could be considered to be
        the only member of their own `match_first` block.
-   If a non-final `match_first` block is chosen, the first matching `impl` in
    that block is selected.

Since it may be observable which `impl` declarations are considered during impl
lookup, due to the [acyclic rule](/docs/design/generics/details.md#acyclic-rule)
and [termination rule](/docs/design/generics/details.md#termination-rule), we
specify this more precisely:

> -   Only (final or non-final)`impl`s with a type structure compatible with the

Author: ```suggestion
-   Only (final or non-final) `impl`s with a type structure compatible with the
```

-   [Proposal #2868](https://github.com/carbon-language/carbon-lang/pull/2868)
    allowed an `impl` to overlap a `final impl` as long as agreed on the
    overlap.

## Proposal

First, to make `require Self impls I` usage more consistent with
`impl Self as I` (and since we will be using this more after this proposal), we
allow the `Self` to be omitted. So:

-   `require Self impls I` ✅ allowed, same meaning as before
-   `require impls I` ✅ allowed, same as `require Self impls I`
-   `require T impls I` ✅ allowed, same meaning as before

Next, since we are creating another way that an `interface` can `extend` another
interface, we change the existing usage of `extend` from being an introducer of
its own declaration, to a modifier on a `require` declaration. In this case, the
only type that can be constrained is `Self`, since that is the type being
extended. To match the restriction that an `extend impl` must be followed by
`as` in `class` scope, `extend require` can only be followed by `impls` in
`interface` scope:

-   `extend require Self impls I` ❌ forbidden, `Self` omitted since no other
    value allowed
-   `extend require impls I` ✅ allowed

> -   `require T impls I` ❌ forbidden

Reviewer 2: Should this have `extend`?

```suggestion
-   `extend require T impls I` ❌ forbidden
```

  let T:! type = U.(J.T);
  fn F() = U.(J.F);
  fn G() = U.(J.G);
}
```

The keyword `final` can be added between `extend` and `impl` to make the
generated `impl` definition `final`. For example, the
`final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T)` from
[the "Problem" section](#problem) would be generated from this:

```carbon
interface ImplicitAs(T:! type) {
  extend final impl as As(T);
}
```

To allow a non-final `impl` to overlap a final `impl`, we say the final `impl`
is prioritized over a non-final `impl` anytime they overlap. This addresses a
[problem](#problem) caused by the fact that we don't support the negative
constraints that would allow a developer to avoid the overlap -- and the only
sound choice is to prefer the final `impl`. However, if a non-final `impl` can
never be used because it is completely subsumed by a final `impl`, that is an
error.

To allow multiple interfaces extending the same interface, we also want to
support a way to declare overlapping final `impl`s. List the names of the
overlapping `impl`s in a `final match_first` block instead of putting `final` on
the `impl` declarations. Note this requires all of the final `impl`s to be in

> the same file.

Reviewer 2: Feels like a slightly awkward transition here...  maybe something like:

```suggestion
To allow multiple interfaces extending the same interface, we also want to
support a way to declare overlapping final `impl`s. We do this by listing the names of the
overlapping `impl`s in a `final match_first` block instead of putting `final` on
the `impl` declarations. Note this requires all of the final `impl`s to be in
the same file.
```

To put the `impl` generated by an `extend impl as Foo` in an interface `Bar` in
a `match_first` block (or a `final match_first` block), we give it the name

> `impl Bar.(as Foo)`.

Reviewer 2: Maybe slightly clearer:

```suggestion
The `impl` generated by an `extend impl as Foo` in an interface `Bar` can be
named as `impl Bar.(as Foo)`. This name can be used to put such an `impl` into 
a `match_first` block (or a `final match_first` block).
```

    [proposal #553](https://github.com/carbon-language/carbon-lang/pull/553).
-   [Proposal #2868](https://github.com/carbon-language/carbon-lang/pull/2868)
    allowed an `impl` to overlap a `final impl` as long as agreed on the
    overlap.

## Proposal

First, to make `require Self impls I` usage more consistent with
`impl Self as I` (and since we will be using this more after this proposal), we
allow the `Self` to be omitted. So:

-   `require Self impls I` ✅ allowed, same meaning as before
-   `require impls I` ✅ allowed, same as `require Self impls I`
-   `require T impls I` ✅ allowed, same meaning as before

Next, since we are creating another way that an `interface` can `extend` another
interface, we change the existing usage of `extend` from being an introducer of
its own declaration, to a modifier on a `require` declaration. In this case, the
only type that can be constrained is `Self`, since that is the type being
extended. To match the restriction that an `extend impl` must be followed by
`as` in `class` scope, `extend require` can only be followed by `impls` in
`interface` scope:

-   `extend require Self impls I` ❌ forbidden, `Self` omitted since no other

>     value allowed

Reviewer 3: ```suggestion
-   `extend require Self impls I` ❌ forbidden, `Self` must be omitted since no other
    value allowed
```
(I was reading this as "`Self` *is* omitted since [...]" which didn't match the example)

  fn Set[addr self: Self*](x: i32) { ... }
  fn Copy[self: Self]() -> Self { ... }
  fn Equal[self: Self](compare: Self) -> bool { ... }
}
```

## Details

Note that `extend` being first, before `final` or `impl`, follows
[our other uses of `extend`](https://github.com/carbon-language/carbon-lang/issues/995#issuecomment-1417391051).

### `extend require`

Some notes and clarifications on `extend require`:

-   To be consistent with other uses of `extend`, we only alias names that don't
    conflict with other names in the interface scope or scopes it extends.
    Example:

    ```carbon
    interface A {
      fn F();
      fn G();
      fn H();
    }

    interface B {
      fn G();
      fn J();
      fn K();
    }

    interface C {
      extend require impls A;
      extend require impls B;
      fn F();
      fn J();
    }
    ```

    This definition of `interface C` is equivalent to:

    ```carbon
    interface C {
      require Self impls A;
      // No `A.G` due to `B.G`
      alias H = A.H;
      require Self impls B;
      // No `B.G` due to `A.G`
      alias K = B.K;
      // Hides `A.F`
      fn F();
      // Hides `B.J`
      fn J();

>     }

Reviewer 3: I think what we want here is that `C.G` is ambiguous, not that it doesn't exist. For example, given
```carbon
interface D {
  fn G();
}
interface E {
  extend require impls C;
  extend require impls D;
}
```
... we don't want `E.G` to unambiguously name `D.G` because `C` has no member named `G` -- instead, that should be ambiguous too.

So I think the rule we want is that we only alias names that don't conflict with other names in the interface scope, but we do alias names that conflict with names in scopes it extends (but any use of such names would be invalid due to ambiguity). Or equivalently, we don't inject aliases, but name lookup into a scope that extends other scopes first looks in that scope, and then if it finds nothing, it looks into all of the extended scopes, and the lookup is an error if there's more than one different result.

Author: Agreed, updated.

    -   Also the current `extend` matches the semantics of `extend` in a named
        constraint.
-   Having a `final impl` of `interface As(T)` currently makes it quite
    difficult to have other impls of the same interface.

On this last point, under the current rules the following example would be
rejected:

```carbon
class C(T:! type) { ... }

interface I {}
impl forall [T:! I] C(T) as As(T) { ... }
```

The issue is that this implementation overlaps with the `final impl` of
`interface As(T)` for any type `C(T)` that implements `ImplicitAs(T)`. This
`impl` is rejected preemptively because it conflicts on the overlap, whether or
not anything inhabits that overlap in the developer's program.

## Background

-   [Proposal #553](https://github.com/carbon-language/carbon-lang/pull/553)
    introduced interface extension
-   [Proposal #983](https://github.com/carbon-language/carbon-lang/pull/983)
    introduced `final impl`
-   [Proposal #2868](https://github.com/carbon-language/carbon-lang/pull/2868)
    allowed an `impl` to overlap a `final impl` as long as agreed on the
    overlap.

## Proposal

> First, since we are creating another way than an `interface` can `extend`

Reviewer 3: ```suggestion
First, since we are creating another way that an `interface` can `extend`
```

another interface, we change the existing usage of `extend` to `extend require`
in both `interface` and `constraint` scopes to distinguish that use. That is,

```carbon
interface A {
  fn F();
}

interface B {
  extend require A;
  fn G();
}
```

is equivalent to:

```carbon
interface A {
  fn F();
}

interface B {
  require Self impls A;
  alias F = A.F;
  fn G();
}
```

> FIXME _**QUESTION**_: Is it `extend require I` or `extend require impls I`?

> FIXME _**QUESTION**_: Should we allow `require impls I` as a shortcut for

> > `require Self impls I` at the same time?

Reviewer 3: It looks like this is switching `extend` from being an introducer in interfaces to always being a modifier. If so, the thing after the modifier ought to be a syntactically valid declaration by itself if we want our syntax to be consistent. So I lean towards saying the syntax is `extend require impls I;` and allowing `require impls I` in general with an implied `Self`. I think it'd probably make sense to also allow `I where impls J` too, meaning `I where .Self impls J`.

Author: I'm fine with that direction. @chandlerc , do you agree?

Author: To avoid surprise in an expression like `I where i32 impls J and impls K` (where it means `.Self impls K` but it might be misread as `J impls K`), perhaps the `impls J` (without a left argument) is only allowed as the first thing after `where`/`require`? `I where impls J and impls K` would not be allowed, but it could be written `I where impls J & K` instead.

Reviewer 3: Another option would be to say that `impls J` is only allowed if it's the *only* thing after `where` / `requires`. That's also the form of the restriction that we want for `extend requires`. But I think I'd be happy either way.

Author: Having talked to @chandlerc , going to go with `require impls`. Doing the same thing for `where` is less clear cut, and will be left out of this proposal; it can be tackled separately if we still want to do it.

Author: Proposal has been updated with this change.

Next, we define the meaning of `extend impl as I;` in an `interface J` scope to
do two things:

> -   Copy the members of `I` to form new members of `J`.

Reviewer 3: To what extent is it observable that we copy the members? From an implementation and teaching standpoint, I think it'd be good if this were equivalent to saying that the members of `I` must also be provided when implementing `J` and that members of `J` can't have names that collide with those of `I`. If we're describing this as copying because that's a convenient approach, I think that's fine, but if this forces the toolchain to actually copy the members and we can't *teach* this as being more like (say) inheritance in C++, I think that'd make me a little unsatisfied.

From an implementation standpoint, I have been imagining that the witness table for `J` will contain a witness for `I` as an element, an impl for `J` builds an impl witness for `I` and populates that entry of `J`'s witness, and the blanket impl for `[forall T:! J] T as I` pulls that witness out of the witness for `T:! J`.

Author: I haven't been worried about the implementation cost since it didn't seem to me like this feature would be used that heavily, but that could very well be wrong. I would be fine with the implementation strategy you suggest, though I'm also fine with copying. It doesn't seem like we can reuse the same witness table since the types of the members are going to change with the change in `Self` type, but maybe you see a path to doing that.

-   Define a blanket `impl` that anything that implements `J` also implements
    `I`, by forwarding to the corresponding members of `J`.

In this example:

```carbon
interface I {
  let T:! type;
  fn F();
  fn G();
}

interface J {
  extend impl as I;
  fn H();
}
```

The definition of `interface J` is equivalent to:

```carbon
interface J {
  let T:! type;
  fn F();
  fn G();
  fn H();
}
impl forall [U:! J] U as I {
  let T:! type = U.(J.T);
  fn F() = U.(J.F);
  fn G() = U.(J.G);
}
```

The keyword `final` can be added between `extend` and `impl` to make the
generated `impl` definition `final`. For example, the
`final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T)` from
[the "Problem" section](#problem) would be generated from this:

```carbon
interface ImplicitAs(T:! type) {
  extend final impl as As(T);
}

> ```

Reviewer 3: How much motivation do we have for the non-`final` form of this? (I guess this allows extending an `interface` from a different file?) That case seems like it might actually require copying of the members, given that `U.(I.T)` and `U.(J.T)` could both be valid but have different values.

Author: At least one case is marking them final using a `match_first` instead when multiple interfaces extend a single one.

sound choice is to prefer the final `impl`. However, if an `impl` can never be
used because it is completely subsumed by a final `impl`, that is an error.

To allow multiple interfaces extending the same interface, we also want to
support a way to declare overlapping final impls. In this case, list the impls
in a `final match_first` block instead of putting `final` on the `impl`
declaration itself. To put the impl generated by an `extend impl as Foo` in an
interface `Bar` in a `match_first` block (or a `final match_first` block), we
give it the name `impl Bar.(as Foo)`.

This example inspired by
[C++ iterator categories](https://en.cppreference.com/w/cpp/iterator) shows
using these pieces together:

```carbon
interface Iterator {
  fn Increment[addr self: Self*]();
}

interface InputIterator {
  let Element:! type;
  extend impl as Iterator;
  fn Get[addr self: Self*]() -> Element;
}

interface OutputIterator {
  let Element:! type;
  extend impl as Iterator;
  fn Set[addr self: Self*](x: Element);
}

// Makes both impls final and prioritizes them.
final match_first {
  impl InputIterator.(as Iterator);
  impl OutputIterator.(as Iterator);

> }

Reviewer 3: Potential problem with this approach: if I have `T:! OutputIterator where .Element = i32`, I know that `T impls Iterator` but I do not know that `T.(Iterator.Element) == i32`. Specifically, if `T` also implements `InputIterator` with a different `.Element`, that `.Element` wins.

Maybe it'd be better to handle this overlap by making all the `extend impl`s final and adding this to the library:
```carbon
interface InputOutputIterator {
  extend final impl as InputIterator;
  extend final impl as OutputIterator;
}
```
... so that a type that is both an input iterator and an output iterator can still have only one element type.

Author: You can create that interface and encourage people to use it. But what if a type implements `InputIterator` and `OutputIterator` without using `InputOutputIterator`? We still need this `final match_first`.

Author: Also, we can't make two interfaces `extend final impl as` a single common interface.

conflict. Further, if there is only one member with that name after the rewrite
in `B`, then there is no way to write the generated `impl` to set the member of
`A` from the values of the associated constants in `B`.

However, if the extending interface gives a value to the associated constant,
there is no need to specify that value in an `impl`, so that is not an error.

```carbon
interface A {
  let T:! type;
  fn F() -> T;
}

interface B {
  let T:! type;
  // Okay
  extend impl as A where .T = T*;
}

class C {
  impl as B {
    // `T` here is `B.T`
    where T = i32;
    // `F` here is `B.F` and `A.F`.
    fn F() -> i32*;
  }
}
```

This is equivalent to:

```carbon
interface A {
  let T:! type;
  fn F() -> T;
}

interface B {
  let T:! type;
  // Skip A.T due to conflict.
  fn F() -> T*;
}

impl forall [U:! B] U as A where .T = .(B.T)* {
  fn F() -> T = U.(B.F);
}

class C {
  impl as B {
    where T = i32;
    fn F() -> i32*;
  }
}
```

This was used in the `OutputForwardIterator` example in
[the "Proposal" section](#proposal).

Name conflicts are not expected to be a problem in practice since this form of

> extension requires the two interfaces to be defined in the same file.

Reviewer 3: Only for `extend final impl`, right?

Author: Added some qualifiers. If we do think this is going to be a problem, though, I'm not sure what to suggest. If we are copying the members automatically we potentially have a problem where names will conflict in the face of evolution. If we are not copying the members automatically, we are doing it manually, which seems even more of a problem since any change will cause it to fall out of date.

    `impl`s in particular it is helpful to be clear about the conditions where
    they won't be selected even when they match.

With this decision not to use type structure to prioritize final `impl`s, it
doesn't make sense to mix `final` and non-final `impl`s in the same
`match_first` block. Furthermore, until the compiler sees the `match_first`
block, we can't treat an `impl` as `final` since we don't know if another final
`impl` will be appear earlier in the `match_first` block and used instead. Which
leads to the decision to put the `final` modifier on the `match_first` block
instead of the `impl` declaration, when it appears in a `match_first` block.
Example:

```carbon
// Compiler does not treat this as final, so it
// won't use `A = i32`.
impl forall [T:! Z] T as J {
  where A = i32;
  // ...
}

// Also not treated as final yet, so the compiler
// won't use `A = bool`.
impl forall [U:! Y] U as J {
  where A = bool;
  // ...
}

// Matching `bool` assignment to `J.A` as the
// previous impl.
impl forall [V:! W] V as J {
  where A = bool;
  // ...
}

final match_first {
  impl forall [U:! Y] U as J;
  impl forall [V:! W] V as J;
  impl forall [T:! Z] T as J;
}

> // Now both are considered final, with the

Reviewer 3: ```suggestion
// Now all are considered final, with the
```

fn G(U:! Y) {
  // Can assume `U impls J` and `U.(J.A)` is `bool`.
}

fn H(V:! W) {
  // Don't know if `V impls Y`, so we don't know
  // which `impl` is selected. We use a symbolic
  // witness for `V` and `V.(J.A)` is unknown
  // (despite that it would be `bool` in either
  // case).
}
```

The compiler will only use a known matching final `impl`'s associated constants

> if it is the first that _could_ match, as determined by type structure. For

Reviewer: ```suggestion
if it is the first that _could_ match, as determined by matching the query type structure against each `impl` in the `match_first`. For
```

I don't know if we have a designated verb for what we do with the type structure - I have been calling it matching, so suggested that here.

The function in the code is `IsCompatibleWith()` so maybe we don't have a verb and we use a relational adjective instead.

```suggestion
if it is the first that _could_ match, as determined by the query type structure being found to be compatible or not with each `impl` in the `match_first`. For
```

Author: I don't think your text makes sense, since only `impl`s have type structure, not queries. I'll try a rephrase.

Note though I suspect the function `IsCompatibleWith` does something different, since we want to consider a `[V:! W] V as L(bool)` query as "could match" `impl forall [U:! Y] C(U) as L(bool)` even though just looking at the type structure it won't consider it matching. Normally every concrete type in the type structure of the impl needs to appear in the query -- we are not normally asking "is there any overlap?".

Also I made this bit its own section.

Reviewer: > I don't think your text makes sense, since only `impl`s have type structure, not queries. I'll try a rephrase.

I guess, in the design. In the implementation, queries are type structures built from the self type and each interface in the facet type query.

> Note though I suspect the function `IsCompatibleWith` does something different, since we want to consider a `[V:! W] V as L(bool)` query as "could match" `impl forall [U:! Y] C(U) as L(bool)` even though just looking at the type structure it won't consider it matching.

These match via `IsCompatiableWith`. The symbolic `V` consumes the `C(` as well as everything to to the matching `)` so they are compatible.

example:

```carbon
interface L(T:! type) {
  let B:! type;
}
class C(T:! type) {}

final match_first {
  // Can use the associated constants from this impl,
  // since it is the first, so nothing could match
  // ahead of it.
  impl forall [U:! Y] C(U) as L(bool) where .B = i32;

  // Can use the associated constants from this impl

>   // only if V can't be `C(U)`, and so it is known

Reviewer: ```suggestion
  // only if query self type can't be `C(U)` (by comparing type structures), and so it is known
```

Author: Rewrote.

  // that it won't use the previous impl instead.
  impl forall [V:! W] V as L(bool) where .B = bool;

  // Can use the associated constants from this impl,

>   // since a query for `T as L(i32)` can't match any
>   // of the earlier impls in the `final match_first`.

Reviewer: ```suggestion
  // since a query type structure that matches this impl will include `L(i32)`, and this type structure won't match any
  // of the earlier impls in the `final match_first`.
```

Author: Rewrote.

fn G(U:! Y) {
  // Can assume `U impls J` and `U.(J.A)` is `bool`.
}

fn H(V:! W) {
  // Don't know if `V impls Y`, so we don't know
  // which `impl` is selected. We use a symbolic
  // witness for `V` and `V.(J.A)` is unknown
  // (despite that it would be `bool` in either
  // case).
}
```

The compiler will only use a known matching final `impl`'s associated constants
if it is the first that _could_ match, as determined by type structure. For
example:

```carbon
interface L(T:! type) {
  let B:! type;
}
class C(T:! type) {}

final match_first {
  // Can use the associated constants from this impl,
  // since it is the first, so nothing could match
  // ahead of it.
  impl forall [U:! Y] C(U) as L(bool) where .B = i32;

  // Can use the associated constants from this impl
  // only if V can't be `C(U)`, and so it is known

>   // that it won't use the previous impl instead.

Reviewer: IIUC this is only known for concrete queries then? As in - any generic in the query self type for `L(bool)` will not receive back a concrete result even though it's from a `final match_first` because there's more than one impl of that interface in the match_first and at least one of them has a generic self?

Author: The query doesn't have to be fully concrete. In this specific case:
```
class B(T:! type) {}
impl forall [X:! type] B(X) as W {}
```
Then `B(X) as W` is true, and so `B(X) as L(bool)` does match 2nd impl, but does not match the type structure of `C(?) as L(bool)` of the first impl, which can be determined even though `B(X)` is not concrete.

Reviewer: I see, this "only if V can't be `C(U)`" part threw me off. I thought it was saying we're comparing the impls against each other in a complex way. What we want is just to compare the type structure of the query against each impl to see if it's a candidate.

Author: Rewrote.

  impl forall [V:! W] V as L(bool) where .B = bool;

  // Can use the associated constants from this impl,
  // since a query for `T as L(i32)` can't match any
  // of the earlier impls in the `final match_first`.
  impl forall [T:! Z] T as L(i32) where .B = f32;
}

class D {}
impl D as W {}
// D as L(bool) definitely uses the middle `impl forall`
// from the `final match_first` so `(D as L(bool)).B` is
// `bool`.
let x: (D as L(bool)).B = true;

class E(T:! type) {}
impl forall [T:! type] C(E(T)) as W {}
// `C(E(T))` does impl `W`, so it impls `L(bool)`. But
// since we don't know whether `E(T)` impls `Y`, we
// don't know which impl it will use, and so

> // `(C(E(T)) as L(bool)).B` is unknown.

Reviewer: As above, this means the result will not be a concrete ImplWitness here, it would have to be a LookupImplWitness even though it landed at a final impl?

Will we be able to diagnose this meaningfully when the user tries to use a concrete thing out of a `final match_first` and we deny them?

Why do we go with a non-concrete result instead of relying on poisoning to prevent the value from changing? Would that not work here due to cross-file concerns?

Author: This is indded an example of no concrete ImplWitness. However, I do not think we should be doing poisoning. Have I left something in the proposal to say that there should be some poisoning involved in this situation?

Reviewer: No, but it's not discussed as an alternative. It would allow use of a final impl to get a concrete result, so that seems like something that could be discussed then, as I find it surprising to select a final impl but not get a concrete ImplWitness back.

We have already said we should poison changing results from generic queries elsewhere: https://github.com/carbon-language/carbon-lang/blob/8c4ff33cb16873effd09768a6916a9d373af0bc7/toolchain/check/testdata/impl/lookup/min_prelude/specialization_poison.carbon#L145-L171

Edit: concrete results -> changing results

Author: The intent is: when you match something in a `final match_first`, but something higher could match if you knew more about the query, you get a `LookupImplWitness` and no poisoning, just like matching an impl that could be specialized. If you match an `impl` in a `final match_first` and no earlier `impl` in the `final match_first` could match, since they all have an incompatible type structure, then you get the witness for that `impl`. Could you suggest edits that would make that clearer (without using `LookupImplWitness`, since that is an internal implementation detail of the toolchain, not something that is part of the design of Carbon as a language).

}
```

> Note: `final match_first` was considered in
> [open discussion on 2023-09-13](https://docs.google.com/document/d/1s3mMCupmuSpWOFJGnvjoElcBIe2aoaysTIdyczvKX84/edit?resourcekey=0-G095Wc3sR6pW1hLJbGgE0g&tab=t.0#heading=h.4r37mjr7c42h)
> as a part of function overloading, and other times since.

This leads us to these rules:

-   An `impl` is considered "final" if is declared with the `final` keyword
    modifier or if it is named in a `final match_first` block.
    -   In the first case (a `final impl` declaration), that `impl` may not
        appear in any `match_first` block.
-   An `impl` may appear in at most one `match_first` block.
    -   This is true whether or not the block is marked `final`.
-   Two final `impl`s in the same file must appear in the same
    `final match_first` block if they overlap, as determined by their type

>     structure.

Reviewer: I had to read back through the proposal carefully to find that we also only allow final impls that overlap in a single file, for now. Maybe we can say that here?

Author: Done.
